motor/discovery/aws/ecs_containers.go

```diff
@@ -0,0 +1,321 @@
+	ecstypes "github.com/aws/aws-sdk-go-v2/service/ecs/types"
+
+	"go.mondoo.com/cnquery/motor/asset"
+	"go.mondoo.com/cnquery/motor/discovery/common"
+	"go.mondoo.com/cnquery/motor/motorid/awsec2"
+	"go.mondoo.com/cnquery/motor/motorid/containerid"
+	"go.mondoo.com/cnquery/motor/platform"
+	"go.mondoo.com/cnquery/motor/providers"
+	aws_provider "go.mondoo.com/cnquery/motor/providers/aws"
+	"go.mondoo.com/cnquery/resources/library/jobpool"
+	"go.mondoo.com/cnquery/stringx"
+)
+
+func NewECSContainersDiscovery(cfg aws.Config) (*ECSContainers, error) {
+	return &ECSContainers{config: cfg}, nil
+}
+
+type ECSContainers struct {
+	config aws.Config
+}
+
+func (ecs *ECSContainers) Name() string {
+	return "AWS ECS Discover"
+}
+
+func (ecs *ECSContainers) List() ([]*asset.Asset, error) {
+	identityResp, err := aws_provider.CheckIam(ecs.config)
+	if err != nil {
+		return nil, err
+	}
+
+	account := *identityResp.Account
+
+	instances := []*asset.Asset{}
+	poolOfJobs := jobpool.CreatePool(ecs.getECSContainers(account), 5)
+	poolOfJobs.Run()
+
+	// check for errors
+	if poolOfJobs.HasErrors() {
+		return nil, poolOfJobs.GetErrors()
+	}
+	// get all the results
+	for i := range poolOfJobs.Jobs {
+		instances = append(instances, poolOfJobs.Jobs[i].Result.([]*asset.Asset)...)
+	}
+
+	return instances, nil
+}
+
+func (ecs *ECSContainers) getRegions() ([]string, error) {
+	regions := []string{}
+
+	ec2svc := ec2.NewFromConfig(ecs.config)
+	ctx := context.Background()
+
+	res, err := ec2svc.DescribeRegions(ctx, &ec2.DescribeRegionsInput{})
+	if err != nil {
+		return regions, err
+	}
+	for _, region := range res.Regions {
+		regions = append(regions, *region.RegionName)
+	}
+	return regions, nil
+}
+
+func (ecs *ECSContainers) getECSContainers(account string) []*jobpool.Job {
+	tasks := make([]*jobpool.Job, 0)
+	var err error
+
+	regions, err := ecs.getRegions()
+	if err != nil {
+		return []*jobpool.Job{{Err: err}} // return the error
+	}
+	log.Debug().Msgf("regions being called for ecs container instances list are: %v", regions)
+	for ri := range regions {
+		region := regions[ri]
+		f := func() (jobpool.JobResult, error) {
+			// get client for region
+			clonedConfig := ecs.config.Copy()
+			clonedConfig.Region = region
+			containerInstancesAssets := []*asset.Asset{}
+			containersAssets := []*asset.Asset{}
+			svc := ecsservice.NewFromConfig(clonedConfig)
+			ctx := context.Background()
+
+			resp, err := svc.ListClusters(ctx, &ecsservice.ListClustersInput{})
+			if err != nil {
+				return nil, errors.Wrap(err, "could not gather ecs cluster information")
+			}
+			for _, cluster := range resp.ClusterArns {
+				containerInstances, err := svc.ListContainerInstances(ctx, &ecsservice.ListContainerInstancesInput{Cluster: &cluster, Status: types.ContainerInstanceStatusActive})
+				if err != nil {
+					return nil, errors.Wrap(err, "could not gather ecs container instances")
+				}
+				if len(containerInstances.ContainerInstanceArns) > 0 {
```

Comment:
I'm not sure if you should discover those as well.
As part of the AWS integration those ec2 instances will already be discovered and scanned. Whether they're part of an ECS cluster or not is not so interesting and will probably already be known from the extracted aws tags.
It also raises the question of how to reconcile the ec2 asset and this one, since it's effectively the same.

Reply:
you're right, they will be discovered

if we detect it's an ec2 instance we use the ec2 instance resolver to add it in, so reconciling won't be a problem

my intention here was so that we can build out a nice map of infrastructure that is being used within ecs in aws. i was thinking future world where we have an aws group that has instances and images and also an ecs group inside of it that holds a reference to the container instance and the containers.... 🤷 @scottford-io would love to hear your opinion on this one

Reply:
@vjeffrey I agree that it is a powerful story to be able to build out a map of the infrastructure as suggested. I believe anything we can do to get users of cnquery more visibility into their infrastructure automatically is something we should always strive for. 

Something to keep in mind is that the person executing cnquery may not be the person who provisioned the infrastructure. I love the idea of this user being able to point cnquery at an AWS account to quickly get a map of what is deployed. 

<img src="https://media0.giphy.com/media/F9DzQnxx6ZZNm/giphy.gif"/>